src/initializers/koa/add-visitor-id.ts

```diff
@@ -1,25 +1,62 @@
 import { Context, Middleware } from 'koa';
 import { getLogger } from 'log4js';
 import { getRequestContext } from '../../builder';
+import { SetOption } from 'cookies';
+import { URL } from 'url';
+import * as uuid from 'uuid';
+import * as ms from 'ms';
 
 const logger = getLogger('orka.visitor');
 
 export const decode = (cookie?: string) => (cookie && JSON.parse(decodeURIComponent(cookie))) || {};
+export const encode = (cookie: any) => encodeURIComponent(JSON.stringify(cookie || {}));
+export const domain = (hostname: string) => {
+  const [_subdomain, _domain, _ext] = (hostname && hostname.split('.')) || [, , , undefined];
+  return (_ext ? ['', _domain, _ext] : _domain ? ['', _subdomain, _domain] : [_subdomain]).join('.');
+};
 
 export default function (config): Middleware {
   return async function addVisitorId(ctx: Context, next: () => void) {
     if (ctx.path === '/health' || !config?.visitor?.cookie || config?.visitor?.enabled === false) {
-      return await next();
+      return next();
     }
     const cookie = ctx.cookies && ctx.cookies.get(config.visitor.cookie);
+    let decoded;
     try {
-      const { cookie_id: visitor } = decode(cookie);
-      ctx.state.visitor = visitor;
-      getRequestContext()?.set('visitor', ctx.state.visitor);
+      decoded = decode(cookie);
+      if (uuid.validate(decoded?.cookie_id)) {
+        ctx.state.visitor = decoded.cookie_id;
+      }
     } catch (e) {
       logger.error(`Failed to parse cookie ${config.visitor.cookie} = ${cookie}`, e);
-    } finally {
-      return await next();
     }
+
+    if (config.visitor.setCookie && !ctx.state.visitor) {
+      ctx.state.visitor = uuid.v4();
+      if (config.visitor.secure && !ctx.cookies.secure) {
+        // Force secure cookies in context, when our ENV dictates to always use secure=true;.
+        ctx.cookies.secure = true;
+      }
+
+      let cookieDomain = domain((ctx.origin && new URL(ctx.origin).hostname) || ctx.hostname);
+      if (config.visitor.getCookieDomain && typeof config.visitor.getCookieDomain === 'function') {
+        cookieDomain = config.visitor.getCookieDomain(ctx);
+      }
+
+      const options: SetOption = {
+        domain: cookieDomain,
+        maxAge: config.visitor.maxAge && ms(config.visitor.maxAge),
```

Comment:
🙃 ms is nice! However maybe we should also document that this is used for someone to know what to add in maxAge